src/python/pants/engine/fs.py

```diff
@@ -248,6 +249,34 @@ class DownloadFile:
     expected_digest: FileDigest
 
 
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class NativeDownloadFile:
+    """Retrieve the contents of a file via an HTTP GET request or directly for local file:// URLs.
+
+    This request is handled directly by the native engine without any additional coercion by plugins,
+    and therefore should only be used in cases where the URL is known to be publicly accessible.
+    Otherwise, callers should use `DownloadFile`.
+
+    The auth_headers are part of this nodes' cache key for memoization (changing a header invalidates
+    prior results) but are not part of the underlying cache key for the local/remote cache (changing
+    a header won't re-download a file if the file was previously downloaded).
+    """
```

Comment:
So, this extra overloading is slightly odd, because as you mention, the "native" implementation already supports multiple URL schemes.

What was the reasoning behind doing this in Python via a union, rather than by adding another implementation on the rust side? A preference for `boto`? A binary size concern? A desire to use the `@union` to plug in other private implementations? Because it seems like it would be simpler to continue to expand the existing scheme based switching on the Rust side (which would get you retry for the auth portion as well), rather than have two code paths parsing URLs to decide which implementation to use.

Reply:
There's a closed PR in your inbox that has this on the rust side... But much less powerful.

Three big things stick out:
- handling auth headers for any URL. As I said in the PR description we might use artifactory, so vanilla http with an auth header is required. That's currently impossible to have rust-side with a flexible enough client-side
- s3 URLs come in many flavors as the S3 tests show. Trying to match them all is folly.
- bloating the engine binary with AWS is just plain rude

Reply:
Would this be clearer if we just made this type private?

Reply:
> Would this be clearer if we just made this type private?

Yea, I would feel better about that. I know that we don't have any explicitly public / stable plugin APIs, but it'd be good not to commit to commit to this being an extension point if we think that support for s3 / auth are things that we should support natively (I think that they are.)

Reply:
To be clear, the native type would be private. The extension point would remain public.

Reply:
How would you use the extension point if the native type was private?

Reply:
Same way I am today, but with an underscore 😌

More honestly, we'd just keep usage internal to the repo I guess

Reply:
Oh maybe there's confusion. Plugin code is currently using DownloadFile, which will become the extension point. Then the extender(?) Is the one using the "native" type.

I have an example usage in this PR for S3, check it out!

Reply:
Oh and maybe you didn't see, the plugin implementation boils down to NativeDownloadFile, so we still get retry and fast-return-if-cached